docs/source/protocol.rst

```diff
@@ -155,108 +154,44 @@ Event ID
 - Listed in :ref:`events page`
 - See: http://linux.die.net/man/3/uuid
 
-Count
-+++++
-
-- 64-bit signed integer
-- GVariant symbol: ``x``
-
-Relative Timestamp
-++++++++++++++++++
-
-- 64-bit signed integer
-- GVariant symbol: ``x``
-- See: http://linux.die.net/man/3/clock_gettime
-
-Auxiliary Payload
-+++++++++++++++++
-
-- Maybe Variant
-- Allows for no contents ``(NULL)`` or content of any type.
-- Used to contain data associated with an event that is logged.
-- GVariant symbol: ``mv``
-- See: https://developer.gnome.org/glib/stable/gvariant-format-strings.html#gvariant-format-strings-maybe-types
-- Details for each event ID listed in :ref:`events page`
-
-Total format
-++++++++++++
-
-In total should look like ``(uayxxmv)``.
-
-Oddities
-++++++++
-
-Aggregates currently have no starting time, only a stopping time. Will be fixed
-by `T8261: Metrics Client: Aggregates record start time <https://phabricator.endlessm.com/T8261>`_
-
-Sequence Metric
-~~~~~~~~~~~~~~~
-
-Sequence metrics are a type of metric that has elements spread across time,
-sequentially (hence the name!) All sequence metrics have a **start** event,
-zero or more **progress** event, and terminate with a **stop** event.
-
-Sequences Metrics' events are chronologically ordered. It should not be
-possible for events in an event sequence to arrive out-of-order
-chronologically.
-
-In order:
-
-User ID
-+++++++
-
-- Unsigned 32-bit integer
-- GVariant symbol ``u``
-- See: http://man7.org/linux/man-pages/man2/getuid.2.html
-
-Event ID
-++++++++
+OS Version
+++++++++++
 
-- Array of 16 unsigned bytes
-- GVariant symbol ``ay``
-- Listed in :ref:`events page`
-- See: http://linux.die.net/man/3/uuid
+- String
+- GVariant symbol: ``s``
 
-Events
+Period
 ++++++
 
-- Array of relative timestamps and auxiliary payloads.
-- First element in array is a start event.
-- Last element in array is a stop event.
-- Any elements in between are progress events.
-- GVariant symbol ``a(xmv)``
+- Unsigned byte
+- GVariant symbol: ``y``
+- Aggregation period (``h`` for hour, ``d`` for day, ``w`` for week, ``m`` for
+  month)
 
-Relative Timestamp
+Absolute Timestamp
 ++++++++++++++++++
 
+- 64 bit signed integer
+- GVariant symbol: ``x``
+- Nanoseconds since the Unix epoch
+- Beginning of the period, with aggregation done using user’s computer time
```

Comment:
I don’t know whether I’m right or wrong, but here’s what I have in mind.

Aggregate events need to use the same aggregation period (same start and stop moments) to be useful in database. For example, daily events will be aggregated for everyone between midnight-UTC and midnight-UTC-day+1. This is needed because we want to group rows by this period, to get sums, counts and/or averages by period.

For example:
- (2021-06-07T00:00:00Z, period 'd') represents 2021-06-07 for daily aggregates,
- (2021-06-07T00:00:00Z, period 'w') represents Week 23 of 2021 for weekly aggregates,
- (2021-06-01T00:00:00Z, period 'm') represents June of 2021 for monthly aggregates.

As the client will do the aggregation, it cannot know for sure when the period starts and stops. It has to rely on its own clock to define when the current period stops and the next period starts. Correcting the period (start and stop moments) on the server side using the timestamp algorithm is possible, but it’s useless: we’ll get different "real" periods for each row, losing the ability to group them by period. It’s not technically impossible to use data with different periods, but it breaks the main interest of having aggregates (which is to group by period, in my opinion).

Sequence events don’t have to be grouped by period, that’s why their start/stop times can be shifted on the server.

Having multiple boots is not the problem here. Having the end of the aggregation period wouldn’t be different from having the period.

As a side note, almost fully unrelated: we group information between two moments (between 2021-06-06T00:00:00Z and 2021-06-07T00:00:00Z, with the Z). We may prefer to group information between two naive datetimes (between 2021-06-06T00:00:00 and 2021-06-07T00:00:00, without the Z) to have periods limits when people are supposed to sleep and have clean variations between week days. If we prefer this, it would be better to use dates (local time, just the date) representation in the protocol instead of Epoch (whose values are different for the same local time in two different timezones), and in the database instead of timezone-aware datetimes.

Reply:
Got it. I think this rationale makes sense. Put another way: we want to be able to summarise across all submitted data by fixed time periods and have that data align reasonably well to those periods. The alignment won't actually be perfect, but close enough is good enough, and we cannot "fix" this on the server-side, as you say.

I'm going back and forth on the naïve versus timezone-aware question… I think it depends on the period. At the level of a week or a month (we should specify whether a week starts on Monday or Sunday…) I don't think it has too big an impact. Perhaps (e.g.) some app usage very early on Monday or late on Sunday is booked against the wrong week; oh well. But at the daily level, I think it may need to be either timezone-aware with machine timezone, or naïve. Otherwise, suppose we have a deployment on the west coast of the USA, expected to be used Monday-Friday (perhaps a school). If we use UTC midnight, that's just 4pm/5pm, and so we would see app/system usage that happens after that point being booked against Saturday. The report would seem "obviously" wrong: the computers are not used on Saturday but metrics say they are.

On the other hand, using local midnight as the cut-off has its own problems: when the timezone changes, the period will not have the correct length. I don't think this is too significant but we'd need to write down the expected behaviour if (e.g.) a computer crosses backwards into a period that has already finished.

Reply:
As discussed, we’ll use the local timestamp. I’ll fix the code and the documentation on metrics-v3 when this is merged.